### DISCUSSION TITLE
use custom NER webservice to set_entities in doc

### DISCUSSION THREAD

**Discussion by aakashb95** [explosion/spaCy#9535] [Help: Coding & Implementations] [answered]
Hi,

I have a webservice for NER that gives me:
(entity, entity_label, (entity_char_idx))
for example:
(John Snow , PERSON, (0,8))

How can i use doc.set_ents to set the entity_label from my webservice to tokens in spacy document? I want to do this so that  i can access them with token.ent_type_ for my downstream tasks.

**Comment by ljvmiranda921** [comment, marked as answer]
Hi @aakashb95 !

You can use `char_span` so that you can slice through character indices. 
Also, the character offsets are incorrect: for `start` it should be the index of the first character, but for `end`, it must be the last character **after** the span:

```python
import spacy

data = [
    ("John Snow", "PERSON", (0, 9)),  # last character after the span
]

nlp = spacy.blank("en")
for text, label, (start, end) in data:
    doc = nlp(text)
    span = doc.char_span(start, end, label=label)
    doc.ents = [span]  
```

Usually, you'd have multiple entities in a given document, that's why we pass a list to `doc.ents`. You can adapt the snippet above to fit your use-case 👍 Also, you can check [the `char_span` docs](https://spacy.io/api/doc#char_span) for more info on the API.

**Reply by aakashb95** [reply]
Thanks for the snippet. I have a follow up question.

Consider:
example = "Mohamed Salah, the 'Egyptian King' scored a hat-trick at Old Trafford. Manchester United defenders had no answer to his movements around the penalty area. On the other hand, Cristiano Ronaldo, the Portuguese superstar, had a goal disallowed. He had very few chances in the game."

I am using [AllenNLP's coreference resolution module](https://demo.allennlp.org/coreference-resolution) to identify mentions that are coreferent to each other:

```python
# pip install allennlp==2.1.0 allennlp-models==2.1.0
from allennlp.predictors.predictor import Predictor
import allennlp_models.tagging
import spacy

nlp = spacy.load("en_core_web_sm")

predictor = Predictor.from_path("https://storage.googleapis.com/allennlp-public-models/coref-spanbert-large-2021.03.10.tar.gz")
res = predictor.predict(example)

clusters = res['clusters']
doc = nlp(example)

all_clusters = []
for cluster_idx in clusters:
    temp_clusters = []
    for coref_idx in cluster_idx:
        start_word_idx, end_word_idx = coref_idx
        mention = doc[start_word_idx : end_word_idx + 1]
        temp_clusters.append(mention)
    all_clusters.append(clusters)
```
Using the above code I get:
```python
all_clusters = [[Mohamed Salah, the 'Egyptian King', his],
 [Cristiano Ronaldo, the Portuguese superstar, He]]
```

the mentions in these clusters are all spacy Spans. 
Now, using my NER webservice, I get the following output:

```python
[('Old Trafford', 'FAC', (57, 69)),
 ('Portuguese', 'NORP', (197, 207)),
 ('Manchester United', 'ORG', (71, 88)),
 ('Mohamed Salah', 'PERSON', (0, 13)),
 ('Cristiano Ronaldo', 'PERSON', (174, 191))]
```

Now I want to use my NER output and using doc.set_ent, and set the entities to those mentions that are inside of coreference clusters. I want to preserve the token word index given by spacy whilst attaching the named entity from my custom webservice.

In your snippet
> ```python
> for text, label, (start, end) in data:
>     doc = nlp(text)
> ```
we're creating doc for the text identified by my NER, whereas I want to set the entities in the original example text. 
Instead of running the webservice again for my downstream tasks later on, I want to be able to see ('Cristiano Ronaldo', 'PERSON') in doc.ents because I manually set them.

How should i use char_span in this scenario? Please let me know if you need any clarifications.

P.S I don't want to use spacy's inbuilt NER and don't have the bandwidth to train spacy NER since we already have a webservice

**Reply by ljvmiranda921** [reply]
Hi @aakashb95 ,

> we're creating doc for the text identified by my NER, whereas I want to set the entities in the original example text.
Instead of running the webservice again for my downstream tasks later on, I want to be able to see ('Cristiano Ronaldo', 'PERSON') in doc.ents because I manually set them.

It's possible to create the `doc` once, or reuse the one you already have. Loop through the entities from the NER service, then attach them into `doc.ents` afterwards:

```python
import spacy

data = [
 ('Old Trafford', 'FAC', (57, 69)),
 ('Portuguese', 'NORP', (197, 207)),
 ('Manchester United', 'ORG', (71, 88)),
 ('Mohamed Salah', 'PERSON', (0, 13)),
 ('Cristiano Ronaldo', 'PERSON', (174, 191))
]

original_text = "...."
nlp = spacy.load("en_core_web_sm")
doc = nlp(original_text)

ents = []
for text, label, (start, end) in data:
    span = doc.char_span(start, end, label=label)
    ents.append(span)
doc.ents = ents
```

If you want to set it on the spans returned by your clusters, it may be a bit tricky since you have to realign your offsets. It may be possible to just do a heuristic based on the length of the word (i.e., `start` is always `0`, and `end` is `len(word)+1`), but may not be foolproof.

**Reply by aakashb95** [reply]
This is great. Thanks. It worked on smaller sentences and articles but failed with E[1010] on a text which had NERs like so:

('G Kishan Reddy', 'PERSON', [1151, 1165]),
 ('Reddy', 'PERSON', [1160, 1165]) 

[E1010] Unable to set entity information for token 223 which is included in more than one span in entities, blocked, missing or outside.

Would you suggest i write a checking utility based on start and end character indexes to eliminate the smaller of the two spans? Is there any other way to bypass this?

**Reply by ljvmiranda921** [reply]
For this, I suggest looking into [various alignment modes in `char_span`](https://spacy.io/api/doc#char_span).  Or most likely, check what's wrong in the annotations.

**Comment by aakashb95** [comment]
I just checked this discussion:
https://github.com/explosion/spaCy/discussions/4966#discussion-60191

How can i incorporate this for my example?

**Reply by ljvmiranda921** [reply]
Hi @aakashb95 ! This follow-up is starting to stray from the original question. For overlapping entities, you can check [SpanGroup](https://spacy.io/api/spangroup#_title). The discussion you linked was from v2, and SpanGroups aren't implemented at that time.

**Reply by aakashb95** [reply]
Sorry, I thought it was a sub-problem of my original question as it dealt with overlapping entities.
You have been super helpful!
If i face issues, i'll raise a different discussion on this specific topic. Thanks!
